### DISCUSSION TITLE
Level of authentication not set after IdP authentication

### DISCUSSION THREAD

**Discussion by blevine** [keycloak/keycloak#25272] [Q&A]
Hi,

Consider the below flow. The Home IDP discovery authenticator delegates authentication to one or more IdPs or locally to Keycloak based on the authenticating user's email domain. When IdP authentication happens, it looks like the level of authentication (LOA) is never set on the authentication session. I realize that the Level 2 OTP authentication in my flow will never happen due to #10250 and I need to figure out how to deal with that. But I'm also guessing that the limitation described in #10250 is causing the LOA not to be set in the authentication session.

When doing IdP authentication, I can see the following TRACE in the log:

> TRACE [org.keycloak.authentication.authenticators.conditional.ConditionalLoaAuthenticator] (default task-1) Condition 'Level 1' evaluated to true due the user not yet reached any authentication level in this session, configuredLoa: 1, requestedLoa: 1

When local (non-IDP) authentication is done, I see the following:

> TRACE [org.keycloak.authentication.authenticators.conditional.ConditionalLoaAuthenticator] (default task-2) Updating LoA to '1' in the condition 'Level 1' when authenticating session 'a8f51bd7-8593-4733-b731-8a863e2e4d8d'. Max age is 36000.
>
>  TRACE [org.keycloak.authentication.authenticators.conditional.ConditionalLoaAuthenticator] (default task-2) Condition 'Level 2' evaluated to false due the requestedLoa '1' smaller than configuredLoa '2'. CurrentAuthenticationLoa: 1
>
>  TRACE [org.keycloak.authentication.authenticators.conditional.ConditionalLoaAuthenticator] (default task-2) Finished authentication at level 1 when authenticating authSession 'a8f51bd7-8593-4733-b731-8a863e2e4d8d'.
>
>  TRACE [org.keycloak.authentication.authenticators.conditional.ConditionalLoaAuthenticator] (default task-2) Updating authenticated levels in authSession 'a8f51bd7-8593-4733-b731-8a863e2e4d8d' to user session note for future authentications: {"1":1701809618}
> 

This is pretty bad. Since the LOA is not set in the authentication session, refreshing the page causes the login prompt to be displayed even though the user has an active session. This is because the CookieAuthenticator contains some logic that checks the LOA and fails if it's not the correct level:
```

    if (acrStore.getRequestedLevelOfAuthentication() > previouslyAuthenticatedLevel) {
        // Step-up authentication, we keep the loa from the existing user session.
        // The cookie alone is not enough and other authentications must follow.
        acrStore.setLevelAuthenticatedToCurrentRequest(previouslyAuthenticatedLevel);
        context.attempted();
```

Is there any workaround for this. I'm not sure that a post IdP login flow could even address this. Please correct me if I'm wrong.
Here's the flow:

![image](https://github.com/keycloak/keycloak/assets/435731/1f899fe5-6db4-45e8-8c15-10b4f683e328)

**Comment by sonOfRa** [comment]
We're currently running into the same bug in a slightly different scenario (we want to treat an IdP as trusted so that users *don't* have to use a second factor when using that specific IdP).

Tried building a kludge-Authenticator that emulates the ConditionalLoaAuthenticator behaviour like this:

```java
/*
 * Force set a specific LoA level. This allows us to emulate a specific LoA to be set by an IdP login such as azure.
 */
public class SetLoaAuthenticator implements Authenticator, AuthenticationFlowCallback {

    private final KeycloakSession session;

    public SetLoaAuthenticator(final KeycloakSession session) {
        this.session = session;
    }

    @Override
    public void authenticate(final AuthenticationFlowContext context) {
        final var authSession = context.getAuthenticationSession();
        final var acrStore = new AcrStore(authSession);
        // This initializes the loa-map into the authSession
        acrStore.setLevelAuthenticated(Integer.parseInt(
                context.getAuthenticatorConfig().getConfig().get(SetLoaAuthenticatorFactory.LOA_CONFIG_KEY)));
        // Write the initialized loa-map to userSession so it is retained for future auth (like #onTopFlowSuccess in ConditionalLoaAuthenticator)
        authSession.setUserSessionNote("loa-map", authSession.getAuthNote("loa-map"));
        context.success();
    }

    @Override
    public void action(final AuthenticationFlowContext context) {}

    @Override
    public boolean requiresUser() {
        return false;
    }

    @Override
    public boolean configuredFor(final KeycloakSession session, final RealmModel realm, final UserModel user) {
        return false;
    }

    @Override
    public void setRequiredActions(final KeycloakSession session, final RealmModel realm, final UserModel user) {}

    @Override
    public void close() {}

    @Override
    public void onParentFlowSuccess(final AuthenticationFlowContext context) {}

    @Override
    public void onTopFlowSuccess() {
        // Write the initialized loa-map to userSession so it is retained for future auth
        final var authSession = session.getContext().getAuthenticationSession();
        authSession.setUserSessionNote("loa-map", authSession.getAuthNote("loa-map"));
    }
}
```

and set it up as a Post Broker Login Flow. Logging in with that IdP on a client that requests a high LoA by default works, but the SSO case (existing session in different client, add session in high loa client) does not work, and a password mask is shown to the user

**Comment by cgeorgilakis** [comment]
Unfortunatelly, the problem will be solved only if [my PR](https://github.com/keycloak/keycloak/pull/12891) for the mentioned issue will be reviewed and accepted.
All other huck solutions are not client specific.
